src/VisualStudio/CSharp/Impl/LanguageService/CSharpCodeCleanupFixer.cs

```diff
@@ -320,37 +400,122 @@ private async Task<bool> FixTextBufferAsync(TextBufferCodeCleanUpScope textBuffe
-                    var isSortUsingsEnabled = context.EnabledFixIds.IsFixIdEnabled(SortImportsFixId);
-                    var enabledDiagnostics = new EnabledDiagnosticOptions(enabedDiagnosticSets.ToImmutableArray(),
-                        new OrganizeUsingsSet(isRemoveUnusedUsingsEnabled, isSortUsingsEnabled));
+                // The project progress tracker will report a TotalItems only up to the number of projects processed so
+                // far, but the total number of completed items will be correct. Assert these conditions in debug
+                // builds, but use bounds checks to ensure we don't throw exceptions at runtime.
+                Debug.Assert(completedItems <= progressTracker.TotalItems);
+                completedItems = Math.Min(progressTracker.CompletedItems, completedItems);
 
-                    var newDoc = await codeCleanupService.CleanupAsync(
-                        document, enabledDiagnostics, progressTracker, cancellationToken).ConfigureAwait(true);
+                Debug.Assert(completedItems - progressTracker.CompletedItems >= 0);
+                var newCompletedItems = Math.Max(0, completedItems - progressTracker.CompletedItems);
 
-                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-                    return document.Project.Solution.Workspace.TryApplyChanges(newDoc.Project.Solution, progressTracker);
+                for (var i = 0; i < newCompletedItems; i++)
+                {
+                    progressTracker.ItemCompleted();
+                }
+            });
+
+            // Prepopulate the solution progress tracker with the total number of documents to process
+            foreach (var projectId in solution.ProjectIds)
+            {
+                var project = solution.GetProject(projectId);
+                if (project.LanguageServices.GetService<ICodeCleanupService>() == null)
+                {
+                    continue;
                 }
+
+                progressTracker.AddItems(project.DocumentIds.Count);
+            }
+
+            foreach (var projectId in solution.ProjectIds)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var project = solution.GetProject(projectId);
+                var newProject = await FixProjectAsync(project, enabledFixIds, projectProgressTracker, cancellationToken).ConfigureAwait(false);
+                solution = newProject.Solution;
             }
+
+            return solution;
+        }
+
+        private async Task<Project> FixProjectAsync(Project project, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            if (project.LanguageServices.GetService<ICodeCleanupService>() == null)
+            {
+                return project;
+            }
+
+            progressTracker.AddItems(project.DocumentIds.Count);
```

Comment:
documentIds.Count seems added twice if call is from FixSolutionAsync

Reply:
It is added twice, but it should not be added twice *to the same progress tracker*. Do you see a case where the value could be added twice to the progress tracker used for Fix Solution progress?

Reply:
This is now simplified and `FixSolutionAsync` doesn't wrap the `ProgressTracker`. Can you review the latest changes?